### PR TITLE
Start work to upgrade to esp-idf 5.0

### PR DESCRIPTION
**Pull request by erikcorry**

### CODE REVIEW

File: src/resources/espnow.cc [hunk 1]
```diff
@@ -150,7 +150,7 @@ static void espnow_send_cb(const uint8_t *mac_addr, esp_now_send_status_t status
   }
 }
 
-static void espnow_recv_cb(const uint8_t *mac_addr, const uint8_t *data, int data_len) {
+static void espnow_recv_cb(const esp_now_recv_info *recv_info, const uint8_t *data, int data_len) {
```
Comment (kasperl) [inline src/resources/espnow.cc:153]:
```suggestion
static void espnow_recv_cb(const esp_now_recv_info* recv_info, const uint8_t* data, int data_len) {
```

File: src/resources/ethernet_esp32.cc [hunk 2]
```diff
@@ -181,6 +184,8 @@ uint32_t EthernetResourceGroup::on_event(Resource* resource, word data, uint32_t
 MODULE_IMPLEMENTATION(ethernet, MODULE_ETHERNET)
 
 PRIMITIVE(init_esp32) {
+  UNIMPLEMENTED();
```
Comment (kasperl) [inline src/resources/ethernet_esp32.cc:187]: Maybe use `UNIMPLEMENTED_PRIMITIVE;`?

File: src/resources/ethernet_esp32.cc [hunk 3]
```diff
@@ -283,10 +288,13 @@ PRIMITIVE(init_esp32) {
   proxy->set_external_address(resource_group);
   return proxy;
 #endif
+#endif
 }
 
 
 PRIMITIVE(init_spi) {
+  UNIMPLEMENTED();
```
Comment (kasperl) [inline src/resources/ethernet_esp32.cc:296]: Maybe use `UNIMPLEMENTED_PRIMITIVE;`?

File: src/resources/uart_esp32_hal.c [hunk 4]
```diff
@@ -100,7 +107,12 @@ void uart_toit_hal_inverse_signal(uart_hal_handle_t hal, uint32_t inv_mask) {
 }
 
 void uart_toit_hal_get_baudrate(uart_hal_handle_t hal, uint32_t *baud_rate) {
```
Comment (kasperl) [inline src/resources/uart_esp32_hal.c:109]:
```suggestion
void uart_toit_hal_get_baudrate(uart_hal_handle_t hal, uint32_t* baud_rate) {
```

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/resources/uart_esp32.cc [hunk 5]
```diff
@@ -287,7 +287,7 @@ UART_ISR_INLINE void RxTxBuffer::read(const SpinLocker& locker, uint8* data, uwo
     memcpy(data + head, buffer, overflow);
     cursor_ = buffer + overflow;
   }
-  available_ -= length;
+  available_ = available_ - length;
```
Comment (floitsch) [inline src/resources/uart_esp32.cc:290]: why?

File: src/rtc_memory_esp32.cc [hunk 6]
```diff
@@ -37,16 +37,7 @@
...
-extern "C" {
-#ifdef CONFIG_IDF_TARGET_ESP32C3
-  #include <esp32c3/clk.h>
-#elif CONFIG_IDF_TARGET_ESP32S3
-  #include <esp32s3/clk.h>
-#else
-  #include <esp32/clk.h>
-#endif
-}
+#include "esp_private/esp_clk.h"
```
Comment (floitsch) [inline src/rtc_memory_esp32.cc:40]: That doesn't look good.
We should look into why we need to reach into the `private` area.
File a bug and/or TODO?

File: src/resources/pulse_counter_esp32.cc [hunk 7]
```diff
@@ -75,7 +77,7 @@ class PcntUnitResource : public Resource {
     // https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/pcnt.html#install-pcnt-channel
     // This static assert might hit, even though the code is still OK. Check the documentation if
     // the code from 'master' (as of 2022-07-01) has already made it into the release you are using.
-    static_assert(ESP_IDF_VERSION_MAJOR == 4 && ESP_IDF_VERSION_MINOR == 4,
+    static_assert(ESP_IDF_VERSION_MAJOR == 5 && ESP_IDF_VERSION_MINOR == 0,
```
Comment (floitsch) [inline src/resources/pulse_counter_esp32.cc:80]: It seems like the new API is available.
I will try to get rid of this static_assert once we have switched.

**Review by floitsch — APPROVED** [review]
LGTM.

**Comment by erikcorry** [comment]
I'm going to keep committing individually reviewed changes to this branch, until we have it working.

File: src/rtc_memory_esp32.cc (hunk 6 above)
Comment (kasperl) [reply src/rtc_memory_esp32.cc]: Fixed.

**Pull request merged**